nucliadb/nucliadb/tasks/consumer.py

```diff
@@ -61,29 +65,49 @@ async def initialize(self, context: ApplicationContext):
         await self._setup_nats_subscription()
         self.initialized = True
 
+    async def finalize(self):
+        self.initialized = False
+        if self.subscription is not None:
+            await self.context.nats_manager.unsubscribe(self.subscription)
+        for task in self.running_tasks:
+            task.cancel()
+        try:
+            await asyncio.wait(self.running_tasks, timeout=5)
+            self.running_tasks.clear()
+        except asyncio.TimeoutError:
+            pass
+
     async def _setup_nats_subscription(self):
         # Nats push consumer
         subject = self.stream.subject
         group = self.stream.group
         stream = self.stream.name
-        await self.context.nats_manager.subscribe(
+        self.subscription = await self.context.nats_manager.subscribe(
             subject=subject,
             queue=group,
             stream=stream,
-            cb=self.subscription_worker,
+            cb=self._subscription_worker_as_task,
             subscription_lost_cb=self._setup_nats_subscription,
+            manual_ack=True,
             config=nats.js.api.ConsumerConfig(
                 deliver_policy=nats.js.api.DeliverPolicy.ALL,
                 ack_policy=nats.js.api.AckPolicy.EXPLICIT,
                 ack_wait=nats_consumer_settings.nats_ack_wait,
                 idle_heartbeat=nats_consumer_settings.nats_idle_heartbeat,
+                max_ack_pending=self.max_concurrent_messages,
             ),
         )
         logger.info(
             f"Subscribed to {subject} on stream {stream}",
             extra={"consumer_name": self.name},
         )
 
+    async def _subscription_worker_as_task(self, msg: Msg):
+        seqid = int(msg.reply.split(".")[5])
+        task_name = f"NatsTaskConsumer({self.name}, msg={seqid})"
+        task = asyncio.create_task(self.subscription_worker(msg), name=task_name)
+        self.running_tasks.append(task)
```

Comment:
Memory leak in `running_tasks`: you're always adding tasks but never removing them. You could add a `add_done_callback` to the task to autoremove itself of check for finished tasks before creating new ones

Reply:
good catch! fixing